pkg/app/ops/insightcollector/deployments.go

```diff
@@ -49,311 +45,86 @@ package insightcollector
-// func extractChangeFailureRate(deployments []*model.Deployment, from, to int64, targetTimestamp int64) (*insight.ChangeFailureRate, []*model.Deployment) {
-// 	var ds []*model.Deployment
-// 	var rest []*model.Deployment
-// 	for _, d := range deployments {
-// 		if d.CompletedAt < to && d.CompletedAt >= from {
-// 			ds = append(ds, d)
-// 		} else {
-// 			rest = append(rest, d)
-// 		}
-// 	}
-// 	var successCount int64
-// 	var failureCount int64
-// 	for _, d := range ds {
-// 		switch d.Status {
-// 		case model.DeploymentStatus_DEPLOYMENT_SUCCESS:
-// 			successCount++
-// 		case model.DeploymentStatus_DEPLOYMENT_FAILURE:
-// 			failureCount++
-// 		}
-// 	}
-
-// 	var changeFailureRate float32
-// 	if successCount+failureCount != 0 {
-// 		changeFailureRate = float32(failureCount) / float32(successCount+failureCount)
-// 	} else {
-// 		changeFailureRate = 0
-// 	}
-
-// 	return &insight.ChangeFailureRate{
-// 		Timestamp:    targetTimestamp,
-// 		Rate:         changeFailureRate,
-// 		SuccessCount: successCount,
-// 		FailureCount: failureCount,
-// 	}, rest
-// }
-
-// // groupDeployments groups deployments by applicationID and projectID
-// func groupDeployments(deployments []*model.Deployment) (apps, projects map[string][]*model.Deployment) {
-// 	apps = make(map[string][]*model.Deployment)
-// 	projects = make(map[string][]*model.Deployment)
-
-// 	for _, d := range deployments {
-// 		apps[d.ApplicationId] = append(apps[d.ApplicationId], d)
-// 		projects[d.ProjectId] = append(projects[d.ProjectId], d)
-// 	}
-// 	return
-// }
+func (c *Collector) collectDevelopmentFrequency(ctx context.Context, ds []*model.Deployment) error {
+	dailyDeployments := groupDeploymentByProjectID(ds)
+
+	for projectID, deployments := range dailyDeployments {
+		err := c.insightstore.Put(ctx, projectID, deployments, deploymentFrequencyMinimumVersion)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// returned deployments are sorted by CompletedAt ASC
+func (c *Collector) findDeploymentsCompletedInRange(ctx context.Context, from, to int64) ([]*model.Deployment, error) {
+	if from > to {
+		return []*model.Deployment{}, nil
+	}
+
+	filters := []datastore.ListFilter{
+		{
+			Field:    "CompletedAt",
+			Operator: datastore.OperatorLessThanOrEqual,
+			Value:    to,
+		},
+	}
+
+	var deployments []*model.Deployment
+	minCompletedAt := from
+	for {
+		d, _, err := c.deploymentStore.List(ctx, datastore.ListOptions{
+			Limit: limit,
+			Filters: append(filters, datastore.ListFilter{
+				Field:    "CompletedAt",
+				Operator: datastore.OperatorGreaterThanOrEqual,
+				Value:    minCompletedAt,
+			}),
+			Orders: []datastore.Order{
+				{
+					Field:     "CompletedAt",
+					Direction: datastore.Asc,
+				},
+				{
+					Field:     "Id",
+					Direction: datastore.Asc,
+				},
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+		if len(d) == 0 {
+			// get all deployments in range
+			break
+		}
```

Comment:
I think it better checking whether being got all deployments or not with cursor.

Reply:
I certainly agree . Thank you!